Bluetooth: Host: Add EATT channel option flag

hermabe opened the pull request:
Adds an option flag to the `bt_gatt_*_params` structs if EATT is enabled so that the application can choose to send a request/indication/notification over enhanced channels, the unenhanced channel or any ATT channel. This is an implementation of the previously posted RFC https://github.com/zephyrproject-rtos/zephyr/pull/41909.

Usages in the tree are updated to use `BT_ATT_CHAN_ANY` which keeps the current behavior.

The option is not added for the write command, as it does not have a parameter struct. We should consider adding this and replacing the current functions that takes the parameters individually.

fixes https://github.com/zephyrproject-rtos/zephyr/issues/41355

hermabe committed:
Bluetooth: Host: Add option to use GATT over EATT

A channel option field has been added to the bt_gatt_*_params structs.
This allows the application to choose wether Unenhanced ATT, Enhanced
ATT, or Any ATT channel shall be used for the
request/command/indication/notification.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe committed:
Bluetooth: Shell: Use chan_opt BT_ATT_CHAN_ANY

The chan_opt field of GATT parameter structs must be set.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe committed:
Bluetooth: Tests: Use chan_opt BT_ATT_CHAN_ANY

The chan_opt field of GATT parameter structs must be set.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe committed:
Bluetooth: Host: Check if channel option is valid

Disallow trying to send requests or notifications on only EATT channels
if no EATT channels are connected or the link is not encrypted. In these
cases the operation will always fail, so it should not be queued.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -1050,6 +1053,9 @@ static inline int bt_gatt_notify(struct bt_conn *conn,
 	params.attr = attr;
 	params.data = data;
 	params.len = len;
+#if defined(CONFIG_BT_EATT)
+	params.chan_option = BT_ATT_CHAN_ANY;
+#endif /* CONFIG_BT_EATT */
```
I think this is probably the best approach to make the change as un-disruptive as possible :)

hermabe replied:
Yes, the behavior should be the same as before. I tried to find all usages in the tree, but I may have missed some.

Thalley reviewed (changes requested):
Looks very promising. A couple of comments. 

Can you also add a `fixes https://github.com/zephyrproject-rtos/zephyr/issues/41355` to the PR description?

hermabe commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2229,10 +2238,9 @@ static void gatt_indicate_rsp(struct bt_conn *conn, uint8_t err,
-					 size_t len)
+static struct bt_att_req *gatt_req_alloc(bt_att_func_t func, void *params, bt_att_encode_t encode,
+					 uint8_t op, size_t len,
+					 enum bt_att_chan_option chan_option)
```
Leftover from a previous iteration of this. Will remove.

hermabe commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3408,9 +3428,9 @@ static int read_included_uuid(struct bt_conn *conn,
-	return gatt_req_send(conn, read_included_uuid_cb, params,
-			     read_included_uuid_encode, BT_ATT_OP_READ_REQ,
-			     sizeof(struct bt_att_read_req));
+	return gatt_req_send(conn, read_included_uuid_cb, params, read_included_uuid_encode,
```
Yes, autoformatting. The `ColumnLimit` is 100 in `.clang-format`, so that is automatically enforced. I don't think clang-format understands "80 columns is recommended".

Thalley replied:
In any case, it is fine that they exceed 80, but it shouldn't automatically move code to always attempt to use the full 100, as that is not only disruptive (modifies lines of code that the PR otherwise doesn't change), which makes it harder to do `git blame`. 

I've seen a lot of issues with people using auto-clang-format, so this isn't the first time I've seen it doing stuff it shouldn't :D The Zephyr (and Linux) code style guide is not properly support for clang format unfortuntately.

hermabe replied:
Ok, will change this to make the diff minimal.

hermabe commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2255,14 +2263,19 @@ static struct bt_att_req *gatt_req_alloc(bt_att_func_t func, void *params,
 
 #ifdef CONFIG_BT_GATT_CLIENT
 static int gatt_req_send(struct bt_conn *conn, bt_att_func_t func, void *params,
-			 bt_att_encode_t encode, uint8_t op, size_t len)
+			 bt_att_encode_t encode, uint8_t op, size_t len,
+			 enum bt_att_chan_option chan_option)
```
At this point we do not know the type of the `params`. Looking up the `chan_option` in the `params` would require the `chan_option` to have the same offset in all the different `bt_gatt_*_params` structs, which we don't want to enforce.

Thalley replied:
The alternative is to do a lookup based on the `func` value, as that can determine the type of `params`, but it may not be worth it.

hermabe commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -981,6 +981,9 @@ struct bt_gatt_notify_params {
 	bt_gatt_complete_func_t func;
 	/** Notification Value callback user data */
 	void *user_data;
+#ifdef CONFIG_BT_EATT
+	enum bt_att_chan_option chan_option;
+#endif /* CONFIG_BT_EATT */
```
It is guarded here to not change the API when EATT is disabled. If we are ok with changing the API also in that case, the guard can be removed.

I made this a bit field after your comment in the RFC (https://github.com/zephyrproject-rtos/zephyr/pull/41909#discussion_r787519694), so that it can more easily be expanded in the future.

Thalley replied:
Yup :) As mentioned, it probably is fine as is, and it's easier to remove the guard in the future than it is change the type of this value.

MariuszSkamra commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -1343,6 +1355,9 @@ struct bt_gatt_discover_params {
 	/** Only for stack-internal use, used for automatic discovery. */
 	struct bt_gatt_subscribe_params *sub_params;
 #endif /* defined(CONFIG_BT_GATT_AUTO_DISCOVER_CCC) */
+#ifdef CONFIG_BT_EATT
+	enum bt_att_chan_option chan_option;
+#endif /* CONFIG_BT_EATT */
```
I am not sure the GATT discovery should be even possible for the user to be done over EATT. It's because the fact the EATT is supported by remote or not is based on the knowledge whether remote has  GATT `SERVER SUPPORTED FEATURES` characteristic value and `EATT Supported` bit is set.

Thus my assumption is that GATT Discovery shall be done over unenhenced bearer only.

BTW, does the implementation reads the `SERVER SUPPORTED FEATURES` before trying to connect EATT channels?

hermabe replied:
It is not required for LE to read the server features before connecting EATT channels and we don't do that currently.

From 6.2 LE GAP INTEROPERABILITY REQUIREMENTS:
> Note: Unlike BR/EDR, it is not necessary to check the Server Supported
Features characteristic before attempting to establish an Enhanced ATT
bearer.

I am not sure how useful discovery over EATT is, but not sure we should disallow it either.

hermabe replied:
@MariuszSkamra ping

MariuszSkamra replied:
OK :)

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -981,6 +981,9 @@ struct bt_gatt_notify_params {
 	/** Notification Value callback user data */
 	void *user_data;
+#ifdef CONFIG_BT_EATT
+	enum bt_att_chan_option chan_option;
```
I think we should add an entry in the release notes about the addition of this: Applications that enable `CONFIG_BT_EATT` will need to ensure that this is correctly set.

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2319,6 +2337,10 @@ static int gatt_indicate(struct bt_conn *conn, uint16_t handle,
 		return -EPERM;
 	}
 
+	if (!bt_att_chan_option_valid(conn, BT_ATT_CHAN_OPTION(params))) {
```
```suggestion
	if (IS_ENABLED(CONFIG_BT_EATT) && !bt_att_chan_option_valid(conn, BT_ATT_CHAN_OPTION(params))) {
```

Thalley reviewed (changes requested):
Minor comments

hermabe committed:
doc: release-notes: Document Bluetooth EATT API change

A field was added to the `bt_gatt_<operation>_params` structs when
`CONFIG_BT_EATT` is enabled. This changes the experimental API for EATT.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe commented on subsys/bluetooth/host/att.c:
```diff
@@ -3613,17 +3631,27 @@ bool bt_att_out_of_sync_sent_on_fixed(struct bt_conn *conn)
+{
+	/* Choosing EATT requires EATT channels connected and encryption enabled */
+	return !((chan_option == BT_ATT_CHAN_ENHANCED) &&
+		 ((bt_conn_get_security(conn) < BT_SECURITY_L2) || bt_att_fixed_chan_only(conn)));
```
This change changed the behavior and made the tests fail. Have rewritten it to be clearer and preserving the behavior of the initial expression.

Thalley reviewed (commented):
PR LGTM now. 

How do we test that it works? i.e. if setting `BT_ATT_CHAN_ENHANCED`, how do we verify that the request is sent over an enhanced ATT?

hermabe commented:
> How do we test that it works? i.e. if setting BT_ATT_CHAN_ENHANCED, how do we verify that the request is sent over an enhanced ATT?

@CarbuneanuAlexandru has added a test PR here:  https://github.com/zephyrproject-rtos/zephyr/pull/45381

There is currently no API for seeing which channel a request/notification/indication was received on. I don't think it is useful for the application either, but it was added just for testing in the above PR.

Thalley commented:
> > How do we test that it works? i.e. if setting BT_ATT_CHAN_ENHANCED, how do we verify that the request is sent over an enhanced ATT?
> 
> @CarbuneanuAlexandru has added a test PR here: #45381
> 
> There is currently no API for seeing which channel a request/notification/indication was received on. I don't think it is useful for the application either, but it was added just for testing in the above PR.

Nice :) 

I assume that the tests are passing with the current implementation. 

I do also not see a use case for the receiver (at application level) to know whether it was received on a enhanced or unenhanced.

Thalley commented on include/zephyr/bluetooth/att.h:
```diff
@@ -91,6 +91,12 @@ size_t bt_eatt_count(struct bt_conn *conn);
 #endif /* CONFIG_BT_EATT */
 
+enum bt_att_chan_option {
+	BT_ATT_CHAN_UNENHANCED = BIT(0),
+	BT_ATT_CHAN_ENHANCED = BIT(1),
+	BT_ATT_CHAN_ANY = BT_ATT_CHAN_UNENHANCED | BT_ATT_CHAN_ENHANCED,
```
Hmm, while reading the test in https://github.com/zephyrproject-rtos/zephyr/pull/45381, this option is actually more disruptive that I though. 

Assuming that the majority of uses of the GATT parameters initially sets all values to 0, the default value for the options is actually invalid (i.e. 0). 

Instead of having
0x00 = invalid
0x01 = ATT only
0x02 = EATT only
0x03 = ATT/EATT

should it perhaps be
0x00 = ATT/EATT
0x01 = ATT only
0x02 = EATT only

In other words, if neither `BT_ATT_CHAN_UNENHANCED` or `BT_ATT_CHAN_ENHANCED`, it will preserve the default behavior of sending on "whatever free channel comes first", without changing the functionality of what this PR is at this point. 

So my suggestion is
```suggestion
	BT_ATT_CHAN_ANY = 0
	BT_ATT_CHAN_UNENHANCED_ONLY = BIT(0),
	BT_ATT_CHAN_ENHANCED_ONLY = BIT(1),
```
and then add a check to verify that both `BT_ATT_CHAN_UNENHANCED_ONLY` and `BT_ATT_CHAN_ENHANCED_ONLY` cannot be set at the same time.

Thalley replied:
Actually `BT_ATT_CHAN_OPT_ANY` should perhaps be `BT_ATT_CHAN_OPT_NONE`

Thalley replied:
I think I'd like to get input from more regarding this, i.e. if `0x00` should retain the current default behavior, or if `0x03` should.

hermabe replied:
Another possibility is for `0x00` and `0x03` both to mean "any".
```c
enum bt_att_chan_option {
        BT_ATT_BACKCOMPAT = 0x00,
	BT_ATT_CHAN_UNENHANCED = BIT(0),
	BT_ATT_CHAN_ENHANCED = BIT(1),
	BT_ATT_CHAN_ANY = BT_ATT_CHAN_UNENHANCED | BT_ATT_CHAN_ENHANCED,
};
```

And update `BT_ATT_CHAN_OPTION` to something like
```c
#define BT_ATT_CHAN_OPTION(_params)  \
    ((_params)->chan_option ? (_params)->chan_option : BT_ATT_CHAN_ANY)
```
I am unsure whether I like it or not.

Thalley replied:
I would rather introduce a breaking change than having a forever-living backcompat change in the API :D

hermabe replied:
I think we should go with your option and have `0x0` be "any". The application should use the named enumerations anyway, only the implementation cares about how the bitfield is laid out.

hermabe commented:
@jhedberg Please review. I would like to get this in before RC1 on Friday.

jhedberg commented on include/zephyr/bluetooth/att.h:
```diff
@@ -91,6 +91,12 @@ size_t bt_eatt_count(struct bt_conn *conn);
 
+enum bt_att_chan_option {
+	BT_ATT_CHAN_ANY = 0x0,
+	BT_ATT_CHAN_UNENHANCED = BIT(0),
```
Is “unenhanced” a term that the spec uses? If not, please use a term from the spec, or if an appropriate one doesn’t exist I’d suggest “legacy”.

hermabe replied:
"Unenhanced" is used in the spec.
> 3.2.11 ATT bearers
Any ATT bearer that is not an Enhanced ATT bearer, using any other L2CAP channel mode, is known as an Unenhanced ATT beare

jhedberg commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -981,6 +981,9 @@ struct bt_gatt_notify_params {
 	/** Notification Value callback user data */
 	void *user_data;
+#ifdef CONFIG_BT_EATT
+	enum bt_att_chan_option chan_option;
```
Wouldn’t “preference” be a better term than “option”? The struct member would then be called eg `chan_ pref`

hermabe replied:
"preference" is a better name for the current use. "option" is better if we want to generalize this field in the future.

jhedberg replied:
If you want to make it completely generic, just call it `flags` or `opts`, however we could also cross that bridge when we get there (if it ever happens) and go with the more intuitive and specific naming for now.

hermabe replied:
Changed to "preference".

jhedberg commented on include/zephyr/bluetooth/att.h:
```diff
@@ -91,6 +91,12 @@ size_t bt_eatt_count(struct bt_conn *conn);
 
 #endif /* CONFIG_BT_EATT */
 
+enum bt_att_chan_option {
```
Is “option” a term that the spec uses? If not, I think “preference” (shortened to “pref”) would be more intuitive.

hermabe replied:
Agree "preference" is more intuitive for the current use of this field.

Thalley replied:
Doesn't `preference` imply that the host may actually use e.g. an unenhanced ATT bearer, even if `BT_ATT_CHAN_ENHANCED` is the only bit set? 

`option` is the term we use for e.g. the advertising options in `bluetooth.h`, and I think the term `option` better describe what this value represent.

jori-nordic replied:
@jhedberg @Thalley do you agree to `prefs` for the naming ?

Thalley replied:
If we want the host to _control_ which type of bearer, then `prefs` is a poor choice. As mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/45332#discussion_r869963284, `prefs` implies that it may be overridden by the stack, which is not only incorrect (as of the current implementation), but also does not the only issue that we have that requires this feature. 

I vote for `options` to be consistent with the options used in `bluetooth.h`.

jori-nordic replied:
I'd be ok with `options`, what do you think @jhedberg ?

jhedberg commented on include/zephyr/bluetooth/att.h:
```diff
@@ -91,6 +91,12 @@ size_t bt_eatt_count(struct bt_conn *conn);
 
 #endif /* CONFIG_BT_EATT */
 
+enum bt_att_chan_preference {
```
Looks much better now, thanks! However, for consistency, please also use the shorter `pref` here, i.e. `bt_att_chan_pref`

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -276,14 +278,29 @@ static int chan_send(struct bt_att_chan *chan, struct net_buf *buf)
+		return true;
+	}
+
+	if (atomic_test_bit(chan->flags, ATT_ENHANCED)) {
+		return (chan_pref == BT_ATT_CHAN_ENHANCED);
+	} else {
+		return (chan_pref == BT_ATT_CHAN_UNENHANCED);
+	}
```
The `chan_pref` (or option) value is a bitfield. So if `chan_pref = BT_ATT_CHAN_ENHANCED | BT_ATT_CHAN_UNENHANCED`, this check fails. Is that intentional?

hermabe replied:
It was intentional (see assert above), as "any" is now `0x0`. 

Do we want to allow both `0x0` and `0x3` to be "any"? Then the bit-field would make sense at the cost of two options having the same meaning.

Thalley replied:
It's a good question. 

The bitfield approach will make it easier to add more fields to the enum without changing existing behavior (in theory at least), but it is kinda funky with the current options. 

Do we have a use case for wanting to send on an unenhanced ATT bearer? If not, then we could consider have `0x00` = "No preference"/"No options", and then only have a single field (for now):
`BT_ATT_CHAN_ENHANCED_ONLY = BIT(0)`

so that we only have an option to say "Enhanced only" which is what we have  use case for. So 0x00 will be any ATT bearer, and 0x01 (`BIT(0)`) will be EATT only.

hermabe replied:
There are some PTS tests that require us to send on unenhanced, though I am not sure if any "real" application needs it. Having a flag only for "EATT only" aligns with the audio use-case, but is less general.

Thalley replied:
> There are some PTS tests that require us to send on unenhanced

That's interesting. So in those tests we are allowed to connect EATT channels, but not allowed to use them? That seems to require either a test spec update or a PTS erratum, rather than an API choice for Zephyr.

hermabe replied:
See eg. GATT/CL/GPM/BV-11-C and friends, which tests client side GATT caching. There the test procedure requires the IUT to first send a request on an unenhanced bearer, then on an enhanced bearer.

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -3613,17 +3635,31 @@ bool bt_att_out_of_sync_sent_on_fixed(struct bt_conn *conn)
+bool bt_att_chan_pref_valid(struct bt_conn *conn, enum bt_att_chan_pref chan_pref)
+{
+	/* Choosing EATT requires EATT channels connected and encryption enabled */
+	if (chan_pref == BT_ATT_CHAN_ENHANCED) {
```
ditto w.r.t `chan_pref` being a bitfield.

Thalley commented on include/zephyr/bluetooth/att.h:
```diff
@@ -91,6 +91,12 @@ size_t bt_eatt_count(struct bt_conn *conn);
 #endif /* CONFIG_BT_EATT */
 
+enum bt_att_chan_pref {
+	BT_ATT_CHAN_ANY = 0x0,
+	BT_ATT_CHAN_UNENHANCED = BIT(0),
+	BT_ATT_CHAN_ENHANCED = BIT(1),
```
Prefer to use the enum name as prefix for the fields:
```suggestion
	BT_ATT_CHAN_PREF_ANY = 0x0,
	BT_ATT_CHAN_PREF_UNENHANCED = BIT(0),
	BT_ATT_CHAN_PREF_ENHANCED = BIT(1),
```

Also consider to rename `ANY` to `NONE` similar to the other options field we use in BT, e.g. `BT_LE_ADV_OPT_NONE`.

Thalley reviewed (changes requested):
Any chance we can merge this PR with the PR with the tests? 
To me it seems that the last update to this PR would have broken things.

hermabe commented:
> Any chance we can merge this PR with the PR with the tests?
To me it seems that the last update to this PR would have broken things.

The renaming will have broken things, but otherwise it should work. I could cherry-pick the test to this PR, but test-specific things can/should be reviewed separately. I don't want to miss RC1 for this API change because of style issues in the test.

hermabe commented:
I would like to get this merged before RC1 tomorrow and there are only two small issues remaining that we need to reach consensus on.

- The name of the channel selection field/enum.
  - We need to choose between "option", "preference", "flags", "opts". I am fine with any of these.
- What values should the different enumerations have?
  - We want the value `0x0` to keep the old behaviour, ie. any ATT channel can be used.
  - We want to be able to require that only enhanced or unenhanced channels are used, so two bits are used for this.
  - Do we allow the value `0x3` to mean "any", the same as `0x0`? I would say yes, as it would be unintuitive behaviour for the bit field otherwise.

@jhedberg @Thalley Please comment on these questions.

Thalley commented:
> I would like to get this merged before RC1 tomorrow and there are only two small issues remaining that we need to reach consensus on.
> 
>     * The name of the channel selection field/enum.
>       
>       * We need to choose between "option", "preference", "flags", "opts". I am fine with any of these.
> 
>     * What values should the different enumerations have?
>       
>       * We want the value `0x0` to keep the old behaviour, ie. any ATT channel can be used.
>       * We want to be able to require that only enhanced or unenhanced channels are used, so two bits are used for this.
>       * Do we allow the value `0x3` to mean "any", the same as `0x0`? I would say yes, as it would be unintuitive behaviour for the bit field otherwise.
> 
> 
> @jhedberg @Thalley Please comment on these questions.

My 2 cents:
We want to _ensure_ that a GATT request is sent on an EATT channel (requirement from the higher layer profile HAP). To me that indicates that this is not a _preference_, as preference indicates something that my be overwritten by e.g. the stack. 

In `bluetooth.h` we have `BT_LE_SCAN_OPT_*`, `BT_LE_ADV_OPT_*`, `BT_LE_PER_ADV_OPT_*`, BT_LE_PER_ADV_SYNC_OPT_*` and `BT_LE_PER_ADV_SYNC_TRANSFER_OPT_*` that are also bitfields of options for a specific instance of something. I think it makes sense to continue that naming scheme. 
In these options we also have a few options that are mutually exclusive, so I personally think that having

```c
	BT_ATT_CHAN_OPT_NONE = 0x0,
	BT_ATT_CHAN_OPT_UNENHANCED_ONLY = BIT(0),
	BT_ATT_CHAN_OPT_ENHANCED_ONLY = BIT(1),
```
Would be the right solution as 
1) 0x00 retains the current behavior
2) We can _ensure_ that something is sent over a specific type of channel (EATT vs ATT)
3) It is expandable. 

Without a use case for ATT-only, it could also be shortened to 
```c
	BT_ATT_CHAN_OPT_NONE = 0x0,
	BT_ATT_CHAN_OPT_ENHANCED_ONLY = BIT(0),
```

hermabe commented:
> Without a use case for ATT-only, it could also be shortened to
>
>	BT_ATT_CHAN_OPT_NONE = 0x0,
>	BT_ATT_CHAN_OPT_ENHANCED_ONLY = BIT(0),

I think this is sufficient for "real" applications. This sidesteps the issue with enhanced and unenhanced bits being mutually exclusive.

The use for testing may be figured out later, potentially having the bit for unenhanced only available during testing.

hermabe commented:
> We can easily document (and verify) that BT_ATT_CHAN_OPT_UNENHANCED_ONLY and BT_ATT_CHAN_OPT_ENHANCED_ONLY are mutually exclusive

This is the current behavior, setting both is not allowed. Not well documented though.

Thalley commented:
> > We can easily document (and verify) that BT_ATT_CHAN_OPT_UNENHANCED_ONLY and BT_ATT_CHAN_OPT_ENHANCED_ONLY are mutually exclusive
> 
> This is the current behavior, setting both is not allowed. Not well documented though.

Right. I also think that the suffix `_ONLY` also makes it clearer that this is an exclusive option.

hermabe commented:
>  I also think that the suffix _ONLY also makes it clearer that this is an exclusive option.

Agree

Vudentz commented:
Guys be warned that if you do add such policy of marking request to only be sent by certain bearer type you will either have to reorder the att->tx_queue or with the current implementation risk adding delays processing it since no channel with the required type may be available the requests needs to be push back until it can be sent blocking the entire queue, the only procedures that don't suffer from these problems are notification and write command since they are not queued one can just have a flag e.g. reliable, which is actually what really matters here since the spec requiring EATT are probably doing it for reliability and not exactly because of being EATT per se, in fact if another bearer were to be introduced with similar guaranties these option would immediately become obsolete.

hermabe commented:
I would like to get this PR going again. We need a decision on the naming of the option field and whether both 0x0 and 0x3 should be allowed for the option field.

The points Luiz brought up are relevant. Do we want a "reliable" option instead of "Enhanced"/"Unenhanced"? Does it make sense/is it useful to have an "unreliable" option? If "unreliable" is not useful, this is similar to the suggestion by Emil above:
> Without a use case for ATT-only, it could also be shortened to
>
> BT_ATT_CHAN_OPT_NONE = 0x0,
> BT_ATT_CHAN_OPT_ENHANCED_ONLY = BIT(0),

Thalley commented:
> I would like to get this PR going again. We need a decision on the naming of the option field and whether both 0x0 and 0x3 should be allowed for the option field.
> 
> The points Luiz brought up are relevant. Do we want a "reliable" option instead of "Enhanced"/"Unenhanced"? Does it make sense/is it useful to have an "unreliable" option? If "unreliable" is not useful, this is similar to the suggestion by Emil above:
> 
> > Without a use case for ATT-only, it could also be shortened to
> > BT_ATT_CHAN_OPT_NONE = 0x0,
> > BT_ATT_CHAN_OPT_ENHANCED_ONLY = BIT(0),

We could use reliable, but I think it may be misleading (as then we'd basically say that using ATT for notifications is unreliable (which it can be considered if the remote device drops notifications from ATT bearers)). 

Basically what we need is a way to request to be sent on ATT (as there is also a PTS test that requires us to sent a request on the ATT channel), EATT or either.

It would be nice if `either == options == 0`, such that this change won't affect any existing devices. 

The only solution I think of that will allow this, is
```c
	BT_ATT_CHAN_OPT_NONE = 0x0,
	BT_ATT_CHAN_OPT_UNENHANCED_ONLY = BIT(0),
	BT_ATT_CHAN_OPT_ENHANCED_ONLY = BIT(1),
```

And then have `BT_ATT_CHAN_OPT_UNENHANCED_ONLY` and `BT_ATT_CHAN_OPT_ENHANCED_ONLY` as mutually exclusive options

Thalley commented:
> > as then we'd basically say that using ATT for notifications is unreliable
> 
> It is unreliable by spec.
> 
> > It would be nice if either == options == 0, such that this change won't affect any existing devices.
> 
> Agree, this is the least disruptive.
> 
> > The only solution I think of that will allow this, is
> > ```
> >   BT_ATT_CHAN_OPT_UNENHANCED_ONLY = BIT(0),
> >   BT_ATT_CHAN_OPT_ENHANCED_ONLY = BIT(1),
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And then have `BT_ATT_CHAN_OPT_UNENHANCED_ONLY` and `BT_ATT_CHAN_OPT_ENHANCED_ONLY` as mutually exclusive options
> 
> This is the current behavior in the PR, but should be enforced and documented better.

What is the next steps for this PR then? More documentation, and then re-review, or do you want re-review now?

hermabe commented:
> What is the next steps for this PR then? More documentation, and then re-review, or do you want re-review now?

I can validate and document the options better.

There is also this issue brought up by Luiz:
> Guys be warned that if you do add such option for marking request to only be sent by certain bearer type you will either have to reorder the att->tx_queue or with the current implementation risk adding delays processing it since no channel with the required type may be available the requests needs to be push back until it can be sent blocking the entire queue, [...]

hermabe commented:
Do we need to add a doxygen group to att.h and an att.rst here: https://github.com/zephyrproject-rtos/zephyr/tree/main/doc/connectivity/bluetooth/api ? Or should the ATT docs simply be added to gatt.rst?

Thalley commented:
> Do we need to add a doxygen group to att.h and an att.rst here: https://github.com/zephyrproject-rtos/zephyr/tree/main/doc/connectivity/bluetooth/api ? Or should the ATT docs simply be added to gatt.rst?

Let's add ATT to the file IMO. 

w.r.t. the comment from Luiz, that is indeed something we need to consider. The TX queue will go from being a FIFO to being a conditional FIFO. I don't think it'll be too bad to make it work. Alternatively we can just create another FIFO for EATT (so that we have a ATT and a EATT FIFO). 

In terms of supporting more bearers in the future: I guess that will be just fine in term of API change, although the GATT/ATT implementation may need to be updated.

hermabe committed:
Bluetooth: Host: Fix bt_eatt_connect returning -EINVAL if conn is NULL

bt_eatt_connect was documented to return -EINVAL if conn is NULL, but
this was not the case. Instead it lead to undefined behavior.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe committed:
doc: Bluetooth: Add ATT section in docs

There was no API reference for the ATT APIs.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

Thalley commented on doc/connectivity/bluetooth/api/att.rst:
```diff
@@ -0,0 +1,9 @@
+.. _bt_att:
+
+Attribute Protocol (ATT)
+############################
```
```suggestion
Attribute Protocol (ATT)
########################
```

hermabe commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -981,6 +981,9 @@ struct bt_gatt_notify_params {
 	/** Notification Value callback user data */
 	void *user_data;
+#ifdef CONFIG_BT_EATT
+	enum bt_att_chan_pref chan_pref;
```
We can change to `options` instead if we can all agree on that.

I was out of office last week, but will push some stuff here soon.

jhedberg replied:
I’m ok changing `pref` to `opt` if you really foresee extending this field for other things in the future.

hermabe committed:
Bluetooth: Host: Move all sending from tx_queue to process_queue

Similar functionality was implemented in two places.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe committed:
Bluetooth: Host: Get first packet matching channel

When sending an ATT packet, do not just try to send the first packet in
the queue, but find the first one that can be sent using the channel.
This makes sure that eg. packets that shall be sent on an unenhanced ATT
channel are not blocked by packets that shall be sent on enhanced ATT
channels.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

jori-nordic commented on subsys/bluetooth/host/att.c:
```diff
@@ -274,14 +276,94 @@ static int chan_send(struct bt_att_chan *chan, struct net_buf *buf)
+
+	__ASSERT_NO_MSG(k_fifo_is_empty(fifo));
+
+	while ((buf = k_fifo_get(&skipped, K_NO_WAIT))) {
```
can you measure the processing time of this re-shuffling of the fifo? say for a 10-el queue on an nRF52 chip.

hermabe replied:
I did some basic measurements using `k_uptime_ticks()` on 52840, queuing up a bunch of notifications. With 14 and 15 elements queued, it seems to take about 8 ticks.

jori-nordic replied:
I guess there can always be a perf PR later if we need it.

jori-nordic commented on subsys/bluetooth/host/att.c:
```diff
@@ -570,38 +652,45 @@ static int bt_att_chan_send(struct bt_att_chan *chan, struct net_buf *buf)
-		if (err >= 0) {
-			break;
+		if (err == -ENOENT && prev &&
+		    (atomic_test_bit(chan->flags, ATT_ENHANCED) ==
```
I don't understand how this condition makes sense. Do you mean to test another channel in the second `test_bit` ?
edit: same comment for `att_req_send_process`

hermabe replied:
Yes, should be `prev->flags` instead of `chan->flags` twice.

hermabe commented on subsys/bluetooth/host/att.c:
```diff
@@ -274,14 +276,94 @@ static int chan_send(struct bt_att_chan *chan, struct net_buf *buf)
 	int err;
 
-	buf = net_buf_get(queue, K_NO_WAIT);
+	buf = get_first_buf_matching_chan(queue, chan);
```
I will add some more testing.

hermabe commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -993,6 +993,9 @@ struct bt_gatt_notify_params {
 	/** Notification Value callback user data */
 	void *user_data;
+#ifdef CONFIG_BT_EATT
+	enum bt_att_chan_opt chan_opt;
```
You mean to ensure that `sizeof(enum bt_att_chan_opt) == 1`? We could do that. I also see we set the `-fshort-enums` for the bsim board and armclang/clang compilers.

Thalley replied:
If we expect `enum`s to have a specific size, then we should pack it. I don't think we need to pack it here though, as the `chan_opt` isn't a value that we need to send over air or anything.

hermabe commented on doc/releases/release-notes-3.1.rst:
```diff
@@ -25,6 +25,9 @@ Changes in this release
 * LoRaWAN: The message type parameter in :c:func:`lorawan_send` was changed
   from ``uint8_t`` to ``enum lorawan_message_type``. If ``0`` was passed for
   unconfirmed message, this has to be changed to ``LORAWAN_MSG_UNCONFIRMED``.
+* Bluetooth: Applications where :kconfig:option:`CONFIG_BT_EATT` is enabled
```
If the struct has static storage, the field will be set to 0 automatically. If not then the field must be set to a valid value. I think most usages of the parameter structs have static storage, as their lifetime must must extend the whole transaction.

Thalley replied:
> `must`? I thought we were having a backwards-compatible field, meaning that `0` was the current implementation.

If the field is indeed `0` (which would be `BT_ATT_CHAN_ANY`) then it has the same behavior as before. But as @hermabe that is only true if the struct is initialized to 0 (e.g. if they are `static`ally declared or `memset`), but that may not always be the case.

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -274,14 +276,94 @@ static int chan_send(struct bt_att_chan *chan, struct net_buf *buf)
 
+static bool att_chan_matches_chan_opt(struct bt_att_chan *chan, enum bt_att_chan_opt chan_opt)
+{
+	__ASSERT_NO_MSG(chan_opt < 0x03);
```
Do we really need to ASSERT or check `enum` values? I would expect the compiler to handle that (should throw a warning/error).

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt_notif/src/client_test.c:
```diff
@@ -153,6 +153,7 @@ static void gatt_discover(void)
 	discover_params.start_handle = BT_ATT_FIRST_ATTRIBUTE_HANDLE;
 	discover_params.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE;
 	discover_params.type = BT_GATT_DISCOVER_PRIMARY;
+	discover_params.chan_opt = BT_ATT_CHAN_ANY;
```
I would actually love to see a test to cover the two options to verify is they work as intended.

hermabe replied:
I am working on a test now.

Thalley replied:
I might be missing something; how is the test actually using the different configurations? It still, to me, look like the tests are still just using `CHAN_OPT_NONE`

hermabe replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/45332/files#diff-09e781ebc5a36992b15e9b3587153a67fe2cfe524c8c0571a69e80ad207ec85bR225. The latest commit adds variations to the existing notify bsim test using different `CHAN_OPT`s.

hermabe commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -993,6 +993,9 @@ struct bt_gatt_notify_params {
 	bt_gatt_complete_func_t func;
 	/** Notification Value callback user data */
 	void *user_data;
+#ifdef CONFIG_BT_EATT
+	enum bt_att_chan_opt chan_opt;
+#endif /* CONFIG_BT_EATT */
```
Didn't want to change the stable API. With the guard, this is only an API change for EATT, which is marked as experimental.

hermabe committed:
Bluetooth: Tests: Add tests for EATT flag API

Checks that both notifications and requests work using the different
channel options.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe commented on include/zephyr/bluetooth/att.h:
```diff
@@ -94,8 +100,25 @@ size_t bt_eatt_count(struct bt_conn *conn);
+ * and both bits may not be set.
+ */
+enum bt_att_chan_opt {
+	/** Both Enhanced and Unenhanced channels can be used  */
+	BT_ATT_CHAN_ANY = 0x0,
```
Can change to `BT_ATT_CHAN_NONE` to be more in line with other option enums, but I read it as "Use no att channel", which is nonsensical. If one reads it as "no option set" it makes more sense.

hermabe replied:
Changed to NONE for consistency.

hermabe commented on include/zephyr/bluetooth/att.h:
```diff
@@ -94,8 +100,25 @@ size_t bt_eatt_count(struct bt_conn *conn);
+	/** Only Unenhanced channels will be used  */
+	BT_ATT_CHAN_UNENHANCED = BIT(0),
+	/** Only Enhanced channels will be used  */
+	BT_ATT_CHAN_ENHANCED = BIT(1),
```
Added the `ONLY` suffix, and `OPT` infix.

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -142,6 +142,7 @@ struct bt_att_tx_meta_data {
 	uint16_t attr_count;
 	bt_gatt_complete_func_t func;
 	void *user_data;
+	enum bt_att_chan_opt chan_opt;
```
It's not a huge deal anyways, so I am fine with leaving it as is.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_notify/test_scripts/_run_test.sh:
```diff
@@ -0,0 +1,38 @@
+ compile it?)\e[39m"
+        exit 1
+    fi
+    timeout 120 $@ &
```
120 seconds seems like a very long timeout

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_notify/test_scripts/gatt_notify_any_any.sh:
```diff
@@ -0,0 +1,10 @@
+# Copyright 2022 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+simulation_id="gatt_notify_any_any" \
```
What's the reasoning behind the "any_any" naming?

hermabe replied:
Combination of which option the server and client uses. There are 4 cases for each of them (OPT_NONE, ENHANCED_ONLY, UNENHANCED_ONLY, mix of both "onlys").

"any" should be renamed to "none".

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_notify/src/gatt_server_test.c:
```diff
@@ -210,10 +267,28 @@ static void test_main(void)
 	{
-		.test_id = "gatt_server",
+		.test_id = "gatt_server_none",
+		.test_post_init_f = test_init,
+		.test_tick_f = test_tick,
+		.test_main_f = test_main_none,
+	},
+	{
+		.test_id = "gatt_server_unenhanced",
+		.test_post_init_f = test_init,
+		.test_tick_f = test_tick,
+		.test_main_f = test_main_unenhanced,
+	},
+	{
+		.test_id = "gatt_server_enhanced",
+		.test_post_init_f = test_init,
+		.test_tick_f = test_tick,
+		.test_main_f = test_main_enhanced,
+	},
+	{
+		.test_id = "gatt_server_mixed",
 		.test_post_init_f = test_init,
 		.test_tick_f = test_tick,
-		.test_main_f = test_main,
+		.test_main_f = test_main_mixed,
```
Btw, it is also possible to provide arguments when starting a test using the `-argstest` argument. It will basically provide the same as this, so there's no need to modify anything, but it could be a way of just having a single main function and a single test_id, by providing the `chan_opt` as a argument to the test.

Thalley replied:
What we have done some of the other places, is that we have a `_<name>.sh` file (which isn't picked up by CI) to run all the tests locally like here: https://github.com/zephyrproject-rtos/zephyr/blob/d4dfb68ee5b083c8db87b9bd5db37041435726b5/tests/bluetooth/bsim_bt/bsim_test_audio/test_scripts/_cap.sh

jori-nordic replied:
alright I was wondering too if it was because of a limitation of CI. I hope we can work out something better than these scripts for twister v2.

hermabe replied:
Yes, that would be a nice improvement over the current setup, but it was postponed until twister v2. I do not what the timeline on that is.

jori-nordic reviewed (approved):
I approve, but please fix this either in this PR on in a subsequent one right after merging it https://github.com/zephyrproject-rtos/zephyr/pull/45332#discussion_r948042423
